clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3560,9 +3612,49 @@ void Sema::MarkDevice(void) {
       }
     }
   }
-  for (const auto &elt : Marker.KernelSet) {
-    if (FunctionDecl *Def = elt->getDefinition())
-      Marker.TraverseStmt(Def->getBody());
+
+  // Previously, we traverse all SYCL functions including kernel functions
+  // and functions called by kernel to do some sema check following same rules.
+  // But this mechanism can't meet our requirements now as different rules may
+  // be required for different type of SYCL kernel. For example, recursion is
+  // not allowed for normal SYCL kernel but recursive function with "noinline"
+  // attribute is permitted in SYCL ESIMD kernel.
+  // Now, we traverse each SYCL kernel together with all functions called by it
+  // directly or indirectly. Before the sema check, we will check kernel type
+  // to decide whether different rules are used. Currently, only normal SYCL
+  // kernel and SYCL ESIMD kernel are supported here. A internal flag in Marker
+  // is used to indicate whether current check is on SYCL ESIMD kernel and all
+  // functions it call directly or indirectly.
+  // The traverse will introduce unnecessary duplicate checks as a function can
+  // be called in different SYCL kernels. For example, if "kernel1" and
+  // "kernel2" both call function "foo" and these 2 kernels are same type, we
+  // firstly check "kernel1" with "foo" and when we check "kernel2", "foo" will
+  // be checked again but it is unnecessary as the rules are same with previous
+  // check. In order to avoid unnecessary checks, we use 2 set to record all
+  // checked functions in different mode, only the one not in set will be
+  // checked.
+  llvm::SmallPtrSet<FunctionDecl *, 10> NonESIMDModeCheckSet;
+  llvm::SmallPtrSet<FunctionDecl *, 10> ESIMDModeCheckSet;
+  for (Decl *D : syclDeviceDecls()) {
+    if (auto SYCLKernelDec = dyn_cast<FunctionDecl>(D)) {
+      if (FunctionDecl *SYCLKernelDef = SYCLKernelDec->getDefinition()) {
+        Marker.SetTraverseEsimdKernel(SYCLKernelDef->hasAttr<SYCLSimdAttr>());
+        auto &SYCLFunctionCheckSet =
+            (SYCLKernelDef->hasAttr<SYCLSimdAttr>() ? ESIMDModeCheckSet
+                                                    : NonESIMDModeCheckSet);
+        Marker.TraverseStmt(SYCLKernelDef->getBody());
+        for (FunctionDecl *SYCLFunctionDec :
```

Comment:
This bit doesn't seem right either.  From this function's perspective, all of this should be managed by the marker it self.